Lib/Src/CUartCom.cpp

```diff
@@ -10,96 +10,231 @@
  * @param uart_de_pin GPIO pin.
  */
-CUartCom::CUartCom(UART_HandleTypeDef *p_huart,
-                   GPIO_TypeDef *uart_de_port,
-                   uint16_t uart_de_pin)
-    : mp_huart(p_huart),
-      m_uart_de_port(uart_de_port),
-      m_uart_de_pin(uart_de_pin)
+bool CUartCom::init(UART_HandleTypeDef *p_huart,
+                    GPIO_TypeDef *uart_de_port,
+                    uint16_t uart_de_pin)
 {
+    bool b_success = true;
     // Check mp_huart is not null
-    if (!mp_huart)
+    if (!p_huart)
+    {
+        Error_Handler();
+        b_success = false;
+        return b_success;
+    }
+    else
     {
-        // Hang application with a toggling ALARM Pin
-        while (1)
+        mp_huart = p_huart;
+        m_uart_de_pin.init(uart_de_port, uart_de_pin);
+    }
+    // Check this hardware is not assigned to another instance
+    for (uint8_t i = 0; i < s_uart_instances; i++)
+    {
+        if (p_huart->Instance == sp_UART[i]->mp_huart->Instance)
         {
-            // TODO: cannot hardcode stuff like this into classes.
-            //            HAL_GPIO_TogglePin(ALARM_GPIO_Port, ALARM_Pin);
-            HAL_Delay(200);
+            b_success = false;
+            return b_success;
         }
     }
+
+    // Assign hardware to instance
+    sp_UART[s_uart_instances] = this;
+    s_uart_instances++;
+
+    return b_success;
+}
+
+/**
+ * @brief Enables UART reception with interrupts
+ * @note UART Interrupts must be enabled in the hardware configuration in order
+ * to use this method
+ *
+ */
+void CUartCom::startRx()
+{
+    HAL_UART_Receive_IT(mp_huart, &m_rx_char, 1);
 }
 
 /**
  * @brief sends message via UART.
  * @param msg Message to send.
+ * @return True if there was space in the transmission queue
  */
-void CUartCom::sendMessage(const std::string &msg)
+void CUartCom::send(const std::string msg)
 {
-    uint16_t msg_len = msg.length();
-
-    // Enable UART_DE Pin
-    if (m_uart_de_port != nullptr)
+    bool event = NON_EMPTY_MSG;
+    // Check event (empty message or not)
+    if (msg.empty())
+    {
+        event = EMPTY_MSG;
+    }
+    // Push msg to queue
+    if (m_tx_queue.size() <= MAX_TX_QUEUE_SIZE && event == NON_EMPTY_MSG)
     {
-        HAL_GPIO_WritePin(m_uart_de_port, m_uart_de_pin, GPIO_PIN_SET);
+        m_tx_queue.push(msg);
     }
 
-    // Convert string to C style
-    const char *c_msg = msg.c_str();
+    // Check state
+    if (mb_state == IDLE)
+    {
+        if (event == EMPTY_MSG)
+        {
+            // Disable interrupt, disable pin
+            __HAL_UART_DISABLE_IT(mp_huart, UART_IT_TXE);
+            m_uart_de_pin.set(false);
+        }
+        else
+        {
+            // Enable USART_DE pin
+            m_uart_de_pin.set(true);
+
+            // Start transmission with interrupts
+            std::string message = m_tx_queue.front();
+            uint8_t msg_length = message.length();
+            strcpy(m_tx_buffer, message.c_str());
```

Comment:
Why do you copy the message out of the buffer? Why not use the buffer as is, and then pop() the queue when transmission is complete?

Reply:
What do you mean copy out of the buffer? I'm retrieving the next string in the queue, then making it a c-type string and copying that into the buffer. I could pop() when tx is complete, but I can't see why.